GSoC - SBML L3V2 models support with some other bug fixes

hemilpanchiwala opened the pull request:
### Major changes made in SBSCL (from GSoC start):
- Before this PR, two PRs have been merged to SBSCL. A complete description of changes made is described in that PRs.
    - PR #40 
    - PR #37 
- Absolute as well as Relative tolerances are now fixed for the simulation of SBML models. This has solved the problems of failing test cases of the SBML Test Suite due to small deviations.
- Now, SBML models with fbc extensions are simulated using FluxBalanceAnalysis in SBMLTestSuiteTest and SBMLTestRunnerWrapper.
- Active objective function variable is added to FluxBalanceAnalysis. And now all the fbc models in the SBML Test Suite pass successfully.
- SBML L3V2 models issues solved:
    - SBML L3V2 models do not have a `fast` property which will be handled now.
    - SBML L3V2 models now can have `constraints`, `kinetic laws`, `events`, `species`, `rules` (assignment rules and rate rules), `initial assignments`, and any others without math. This has been handled now through this PR in SBSCL.
    - Logical operators are now added to compileDouble in ASTNodeValue.
    - New functions are added to the ASTNodeInterpreter class. (issue #42)
    - Reaction without reactants and products has been solved now. (issue #41)
    - `RELATIONAL_GT`, `RELATIONAL_GEQ`, `RELATIONAL_EQ`, `RELATIONAL_NEQ`, `RELATIONAL_LEQ`, and `RELATIONAL_LT` are now updated to have multiple children.
    - SBSCL now supports `rateOf` csymbol that is newly added in SBML L3V2.
    - New variables have been added to `RateRuleValue` and `RuleValue` classes.
    - A new array `changeRate` has been added to SBMLinterpreter class that keeps the value of the derivatives at the current time.
    - `constantHash` has also been added to SBMLinterpreter that has key as variables and values are boolean (true if constant else false).
    - After making this changes, about 67 SBML L3V2 tests fail excluding the comp tests from the SBML Test Suite. 
- Many other small issues have been resolved in this PR.

This PR closes #38, closes #41, closes #42, closes #43.
More information about changes made till now in SBSCL can be found at my blog posts [[Link](https://gsoc20hemil.blogspot.com/)].

matthiaskoenig commented on src/main/java/org/simulator/sbml/SBMLinterpreter.java:
```diff
@@ -69,25 +70,7 @@
-import org.simulator.sbml.astnode.SpeciesValue;
-import org.simulator.sbml.astnode.StoichiometryValue;
-import org.simulator.sbml.astnode.TimesValue;
+import org.simulator.sbml.astnode.*;
```
Don't do `*` imports. This is normally not very good style and clutters the namespace.

hemilpanchiwala replied:
Updated

hemilpanchiwala commented:
> You have a ton of formating changes of comments in this pull request which makes it very hard to review the actual changes. In the future do separate pull requests for formating changes (of the 98 changed files only ~5-6 were actually changed)

@matthiaskoenig, will take care of it from next time.

hemilpanchiwala commented on src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java:
```diff
@@ -52,7 +55,7 @@ public static void main(String[] args) throws IOException, XMLStreamException, M
         String outputDirPath = args[2];
         String level = args[3];
-        String  version = args[4];
+        String version = args[4];
```
@matthiaskoenig, I am storing in amountHash as SBMLinterpreter takes amountHash as a parameter of its constructor.

hemilpanchiwala commented on src/test/scripts/sbml_test_suite_runner_wrapper.sh:
```diff
@@ -9,4 +9,4 @@
 DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
-java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/* org.simulator.examples.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
+java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/*:$DIR/../../../src/lib/maven/kisao/libkisao/1.0.3.1-rc/libkisao-1.0.3.1-rc.jar:$DIR/../../../src/lib/maven/scpsolver/GLPKSolverPack/4.35v2/GLPKSolverPack-4.35v2.jar:$DIR/../../../src/lib/maven/scpsolver/LPSOLVESolverPack/5.5.2.5/LPSOLVESolverPack-5.5.2.5.jar:$DIR/../../../src/lib/maven/scpsolver/SCPSolver/1.0v2 ...
```
@matthiaskoenig, These new libraries I stated were already present in the SBSCL directory and are used for simulating FBC models, so I added their path explicitly. 

For making a general import I will have to copy the same libraries to the folder `sbml_test_runner_wrapper`.

hemilpanchiwala commented on src/main/java/org/simulator/math/MaxDivergenceTolerance.java:
```diff
@@ -6,102 +6,95 @@
- *   </mfrac>
- *   <mo>&#x2264;<!-- ≤ --></mo>
- *   <mn>1</mn>
+ * <mfrac>
```
@matthiaskoenig, I have made a formatting change according to the JSBML templates. I will update it.

hemilpanchiwala replied:
Updated

hemilpanchiwala commented on src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java:
```diff
@@ -5,16 +5,17 @@
 import org.sbml.jsbml.SBMLDocument;
```
Updated

zakharc commented on src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:
```diff
@@ -0,0 +1,237 @@
+                CompSimulator compSimulator = new CompSimulator(sbmlfile);
+                double stepSize = (duration / steps);
+
+                solution = compSimulator.solve(stepSize, duration);
```
If we will not provide any changes to the CompSimulator one needs to deal with exceptions here (we are throwing except. in the corresponding solve method). Please consider this.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -37,30 +39,33 @@
 
     @Before
-    public void setUp(){ }
+    public void setUp() {
+    }
 
     @After
-    public void tearDown(){ }
+    public void tearDown() {
+    }
```
Do we need to override these methods if we do nothing? As far as I can remember, JUnitTestRunner does nothing if there is no @Before, @After, @SetUp, or @TearDown methods are specified.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                    try {
+                        compSimulator = new CompSimulator(sbmlFile);
+                    } catch (Exception e) {
+                        errorInCompSimulator = true;
```
This is not good. Why we would suppress the exception in the test? That is why we write the tests. If everything is working - we are satisfied, if not - we want to know the reason. Please feel free to throw an exception in the tests. JUnit will treat this as an error and you will be able to see the stack trace to solve the problem fast and not spend your time guessing why this problem occurs.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                    Assert.assertNotNull(compSimulator);
+                    Assert.assertFalse(errorInCompSimulator);
+
+                    //solve
```
I am not sure if this comment necessary here or explains something.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                        double stepSize = (duration / steps);
+                        solution = compSimulator.solve(stepSize, duration);
+                    } catch (Exception e) {
+                        errorInSolve = true;
```
Please see related comment above.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                    try {
+                        solver = new FluxBalanceAnalysis(org.sbml.jsbml.SBMLReader.read(sbmlFile));
+                    } catch (Exception e) {
+                        errorInFBASimulator = true;
```
Please see related comment above.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                    FluxBalanceAnalysis solver = null;
+                    boolean errorInFBASimulator = false;
+                    try {
+                        solver = new FluxBalanceAnalysis(org.sbml.jsbml.SBMLReader.read(sbmlFile));
```
Just curious: why we use the full canonical name here? Do we bind the corresponding library dynamically?

hemilpanchiwala replied:
@zakharc, as there is an another import used `import org.sbml.jsbml.xml.stax.SBMLReader` that's why it is using full canonical name.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+                    try {
+                        solver.solve();
+                    } catch (Exception e) {
+                        errorInSolve = true;
```
Please see related comment above.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+
+                    for (Map.Entry<String, Double> mapElement : inputSolution.entrySet()) {
+                        if (solution.containsKey(mapElement.getKey())) {
+                            Assert.assertEquals(mapElement.getValue(), solution.get(mapElement.getKey()), 0.0002);
```
Whatever this magical number means (0.0002) - I would exclude it to a global var.

zakharc commented on src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
+
+                    Map<String, Double> solution = solver.getSolution();
+
+                    System.out.println(solution);
```
Please see related comment above.

zakharc commented on src/test/scripts/sbml_test_suite_runner_wrapper.sh:
```diff
@@ -9,4 +9,4 @@
 DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
-java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/* org.simulator.examples.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
+java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/*:$DIR/../../../src/lib/maven/kisao/libkisao/1.0.3.1-rc/libkisao-1.0.3.1-rc.jar:$DIR/../../../src/lib/maven/scpsolver/GLPKSolverPack/4.35v2/GLPKSolverPack-4.35v2.jar:$DIR/../../../src/lib/maven/scpsolver/LPSOLVESolverPack/5.5.2.5/LPSOLVESolverPack-5.5.2.5.jar:$DIR/../../../src/lib/maven/scpsolver/SCPSolver/1.0v2 ...
```
This looks scary :)

Can we deal with this in a more generic way? For example: exporting system global variables and running this command with them as arguments.

hemilpanchiwala replied:
I just added the needed jars in the sbml_test_runner_wrapper folder which made things cleaner.

hemilpanchiwala commented:
> @hemilpanchiwala: Some of the comments are not related to your code. Please address the comments related to your changes first. For the rest: I am OK if you will adjust the rest during the next phase.
> 
> Small comments for the future:
> 
> * Please always exclude the formatting changes to the separate PR. I've seen a lot of whitespace changes/etc. in this PR. This will only add work for the reviewer and is not interesting at all.
> * I would suggest looking into coding conventions and guidelines of Eclipse Foundation (if you didn't do this already): https://wiki.eclipse.org/Development_Conventions_and_Guidelines
> 
> Nice work.

@zakharc, Extremely sorry for adding the formatting in this PR. I will take care from next time. And, I have formatted the code according to the template provided by the JSBML community linked [here](https://github.com/sbmlteam/jsbml/tree/master/dev).

hemilpanchiwala commented on src/main/java/org/simulator/comp/CompSimulator.java:
```diff
@@ -16,78 +16,80 @@
+    return doc;
+  }
+
+  public SBMLDocument getDocFlat() {
```
@zakharc, or maybe getFlattenedDoc?

zakharc replied:
Sure. Nice idea.

hemilpanchiwala commented on src/main/java/org/simulator/plot/PlotProcessedSedmlResults.java:
```diff
@@ -37,117 +37,102 @@
\ No newline at end of file
+  /**
+   * Get absolute parent path for given test resource.
+   * Due to the relative paths of SBML and SED-ML files the resource loading is not working
+   * in maven.
+   * <p>
```
@zakharc, it gives example of how to input the parameter.

zakharc replied:
The method signature implies that we want to get a folder path for a test resource and return a string. The javadoc explains that we get an absolute parent path for a given test resource. The "Example: ..."  gives an example of how to input the parameter. But the comment "Due to the relative paths of SBML and SED-ML files the resource loading is not working in maven." has no relation to the method at all.

If we want to explicitly show, that the relative path is awaited as a parameter I would simply explain this in the corresponding @param description of javadoc.

hemilpanchiwala commented on src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:
```diff
@@ -0,0 +1,237 @@
+     */
+    public static void main(String[] args) throws IOException, XMLStreamException, ModelOverdeterminedException, DerivativeException {
+
+        // Configuration
+        String dirPath = args[0];  // path of sbml-test-suite
+        String currentCase = args[1];  // current test case
+        String outputDirPath = args[2];  // output directory for results
+        String level = args[3];  // SBML level
+        String version = args[4]; // SBML version
+
+        String filePath = dirPath + File.separator + currentCase + File.separator + currentCase + "-sbml-l" + level + 'v' + version + ".xml";
+        String settingsPath = dirPath + File.separator + currentCase + File.separator + currentCase + "-settings.txt";
+        String outputFilePath = outputDirPath + File.separator + currentCase + ".csv";
+        String resultsPath = dirPath + File.separator + currentCase + File.separator + currentCase + "-results.csv";
+
+        Properties properties = new Properties();
+        properties.load(new BufferedReader(new FileReader(settingsPath)));
+        double duration;
+        double steps = (!properties.getProperty(STEPS).isEmpty()) ? Double.parseDouble(properties.getProperty(STEPS)) : 0d;
+        Map<String, Boolean> amountHash = new HashMap<String, Boolean>();
+        String[] amounts = String.valueOf(properties.getProperty(AMOUNT)).split(",");
+        String[] concentrations = String.valueOf(
+                properties.getProperty(CONCENTRATION)).split(",");
+        double absolute = (!properties.getProperty(ABSOLUTE).isEmpty()) ? Double.parseDouble(properties.getProperty(ABSOLUTE)) : 0d;
+        double relative = (!properties.getProperty(RELATIVE).isEmpty()) ? Double.parseDouble(properties.getProperty(RELATIVE)) : 0d;
+
+        for (String s : amounts) {
+            s = s.trim();
+            if (!s.isEmpty()) {
+                amountHash.put(s, true);
+            }
+        }
+
+        for (String s : concentrations) {
+            s = s.trim();
+            if (!s.isEmpty()) {
+                amountHash.put(s, false);
+            }
+        }
```
@zakharc, this class is not for testing but instead is used for getting the results in the SBML Test Runner. But yes, I will make this class more cleaner by diving the code into multiple methods.

zakharc replied:
I was implying, that this piece of code could go to a separate object. Let's say SBMLTestRunnerSuiteConfiguration. In this way, we can encapsulate configuration initialization in a separate class. As a result, we can test this with a separate component test. Let's say SBMLTestRunnerSuiteConfigurationTest. It is an idea for the future refactoring, that you might want to do.

But on this stage, I would suggest dealing with failing tests first. So I will mark this as resolved.

hemilpanchiwala commented on src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:
```diff
@@ -0,0 +1,237 @@
+
+        // writes results to the output file in CSV format
+        File outputFile = new File(outputFilePath);
+        outputFile.createNewFile();
```
@zakharc, No, it doesn't need exception as it just deletes an old file (if it already exists) and add new one to the desired location.

zakharc replied:
@hemilpanchiwala: The procedure itself can be not allowed under certain circumstances. Example: 

- We want to delete the file (as it just deletes an old file) and this is not allowed due to the changed writing rights (chmod 555 on linux etc.). 
- The CI system executes your code as a third-person user and has no credentials to do this.

For me, it looks like this should result in an exception. I am not sure what API's we are using "under the hood" (java.io.File, java.nio.file.Paths, ore something else). But to my understanding - every single one of them will result in an exception.

hemilpanchiwala commented:
@zakharc, I have updated most of the requested changes. I will soon update the SBMLTestSuiteRunnerWrapper class by dividing the code into different methods. And for the JavaDoc comments of the overridden methods and some others, I will update it as I work. However, all comment changes pointed by you have been updated.

shalinshah1993 commented:
Lots of good work here! 

- For the white space diffs, it seems that they can be hidden away by clicking on settings and avoid white space changes. However, to avoid pushing them in commits in future, you need to change git attributes. Please add these lines in the `.gitattributes` file. It is the default template for Java code to alway convert the code to and will avoid the white space diffs. You can find default template for other languages here - https://github.com/github/gitignore 

```
# avoid
text=auto 
eol=lf

# Java sources
*.java          text diff=java
*.gradle        text diff=java
*.gradle.kts    text diff=java

# These files are text and should be normalized (Convert crlf => lf)
*.css           text diff=css
*.df            text
*.htm           text diff=html
*.html          text diff=html
*.js            text
*.jsp           text
*.jspf          text
*.jspx          text
*.properties    text
*.tld           text
*.tag           text
*.tagx          text
*.xml           text

# These files are binary and should be left untouched
# (binary is a macro for -text -diff)
*.class         binary
*.dll           binary
*.ear           binary
*.jar           binary
*.so            binary
*.war           binary
*.jks           binary
```

- I saw some coding convention pointed out up here, while I agree it is a good practice to adhere to them, I was thinking may be we can incorporate style checks. For python, I know `pylint` can be easily incorporated with Jenkins. A quick search for java showed checkstyle. https://checkstyle.sourceforge.io/ @draeger how the timeline for this project, may be if it is really easy, we can incorporate lint checker with maven and travis? 

- I also see that that one test is failing. Before this gets merged, can you either fix the test or for now comment it out so it doesn't break the `master` branch?

hemilpanchiwala commented:
@shalinshah1993, I will add the snippet you gave in the `.gitattributes` file in the next commit. But for the tests, I don't think there is a single failing test instead there are many SBML Test Suite tests that are failing currently due to some SBML L3V2 and other issues.

hemilpanchiwala commented:
> Why did subclassing not work? I.e. only overwriting the `solve` method?
@matthiaskoenig,
It did not work as the new instance of GlpkSolver was made in the solve() method explicitly. We did not have any reference to that class in SBSCL. 

Here, the thing is going on as follows:
We make an object of LinearProgramSolver (which is an interface) that is implemented by GLPKSolver class where the solve() method implementation is found. In this solve() method, scpsolver creates an instance of the GlpkSolver class which was creating all the problems and so making that instance (of GlpkSolver) null using deleteProb() solves the issue. But this is all in scpsolver so I made a copy of the GLPKSolver class with making the instance of GlpkSolver null on completing its usage. Once scpsolver makes this change then we can remove this class.

**Note:** GLPKSolver and GlpkSolver are different classes in scpsolver.

matthiaskoenig commented on src/main/java/org/simulator/examples/SimulatorExample.java:
```diff
@@ -31,78 +31,96 @@
-    if (solver instanceof AbstractDESSolver) {
-      ((AbstractDESSolver) solver).setIncludeIntermediates(false);
-    }
+        String fileName = "";
```
Normally one initializes with `null`. Empty string is always a very risky choice. Change to `null`.

hemilpanchiwala replied:
Updated the code!

matthiaskoenig commented on src/main/java/org/simulator/fba/FluxBalanceAnalysis.java:
```diff
@@ -50,433 +50,419 @@
+   * @return
+   */
+  public String getActObjFunc() {
+    return actObjFunc;
```
Better name is `activeObjective`. Always write words out. There is code completion which makes larger names easy to use. Also rename functions to `getActiveObjective`.

hemilpanchiwala replied:
Updated!

matthiaskoenig commented:
Will merge this as soon as the minor changes & the test case are there.

hemilpanchiwala commented:
> Please add a FluxBalanceAnalysis test which:
> 
> * takes the ecoli core model, simulates a FBA, compares objective; currently a simple test case for FBA outside of the SBMLTestSuite is missing

@matthiaskoenig, I have updated the changes. And FBA test was there but it was named CobraSolverTest and was ignored by @Ignore tag. I have renamed it to FluxBalanceAnalysisTest and also removed the ignore tag.
